### PR TITLE
Delete `GT_ASG`

### LINKED ISSUE #10873: Remove GT_ASG nodes
This came up occasionally in discussions but AFAIK there's no issue associated with it and IMO there should be due to the significant (negative) impact `GT_ASG` nodes have on RyuJIT's IR. 

IMO it's one of the worst IR "features", if not plain and simple the worst. So far I only encountered pros and no cons:
* It's an extra IR node - wastes memory and slows down IR traversal.
* It imposes a model where the meaning of some nodes depends on the context - a `GT_LCL_VAR` on the RHS is a use but on the LHS is a def. The JIT uses various means to deal with this issue - `GTF_VAR_DEF`, `GTF_IND_ASG_LHS`, `gtGetParent()` etc.
* It complicates UD chain traversal. SSA gives you the `GT_LCL_VAR` node on the LHS, from there you need to use `gtGetParent()` to get the `GT_ASG` node and then get the assignment's second operand. If you're lucky the LHS follow the assignment, otherwise `gtGetParent()` will need to traverse multiple nodes.
* It requires additional work to transform HIR into LIR, the later does not use assignment nodes.
* It doesn't match the IL model that uses stores. While I don't think that the IR has to strictly follow the IL model this one seems like a huge and unnecessary deviation.
* It doesn't follow the assembly model either. One could argue that `mov eax, ebx` is assignment but that's only true if `eax` and `ebx` happen to be enregistered variables. Otherwise you'd be looking at `mov [eax], ebx` which is more similar to a store.

Anyone knows any advantages?

AFAIR Phoenix did have `Assign` but its IR was very different, looking more like assembly. I know next to nothing about LLVM but I don't think it has assignment.

category:implementation
theme:ir
skill-level:expert
cost:extra-large

### PR DESCRIPTION
**Pull request by SingleAccretion**
After many years of existing and some weeks of removal work, `GT_ASG`, the last large IR quirk, becomes no more, giving way to the store nodes in the entirety of the compiler.

Closes #10873.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=274619&view=ms.vss-build-web.run-extensions-tab):

1) One regression in tests because a SIMD local does not get promoted due to the changed order of how structs are handled in the importer - cases with return buffers used to not go throw `gtNewAssignNode`, thus no "is used as a SIMD intrinsic" flag was set for them. Arguably it still shouldn't be, but at just one diff it did not seem worthwhile to spend code handling this given that promotable SIMD locals are rare.
2) A correctness fix in tests where `GTF_IND_VOLATILE` is now not lost for a struct store.

Of course, there is also a very nice TP diff.

**Comment by BruceForstall** [comment]
Might as well run some stress while waiting for code reviews...

File: src/coreclr/jit/fginline.cpp
Comment (EgorBo) [inline src/coreclr/jit/fginline.cpp:481]: Same as tree->OperIsLocalStore right?

Comment (SingleAccretion) [reply src/coreclr/jit/fginline.cpp:481]: No, `OperIsLocalStore` also includes `STORE_LCL_FLD`, though we won't see it here (for now). This branch only dealt with `ASG(LCL_VAR, ...)`.

**Comment by EgorBo** [comment]
This a huge work requiring deep understanding of all pieces in JIT, Awesome!!

Regarding the way the new node is displayed in dumps, e.g.:
```
                    //       ├──▌  STORE_LCL_VAR int    V35 tmp31
                    //       │  └──▌  LCL_FLD   int    V03 loc1         [+4]
```
Does it look good for everyone? @dotnet/jit-contrib 
Previously we had:
```
                    //       ├──▌  ASG       int
                    //       │  ├──▌  LCL_VAR   int    V35 tmp31
                    //       │  └──▌  LCL_FLD   int    V03 loc1         [+4]
```

File: src/coreclr/jit/promotiondecomposition.cpp
Comment (BruceForstall) [inline src/coreclr/jit/promotiondecomposition.cpp:744]: Here and elsewhere in this function, we previously created a new node to pass to `AddStatement()`. Now, we pass `m_store` directly. Is that correct? Do we need to `gtClone(m_store)`? I don't understand the usage of `m_store`, but I want to ensure we're not linking a node in to the IR that is also used elsewhere. cc @jakobbotsch

**Review by BruceForstall — COMMENTED** [review]
Really nice!

Someday we should probably also update https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/ryujit-overview.md and https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/ryujit-tutorial.md (maybe more?).

**Comment by SingleAccretion** [comment]
Stress results analysis:

1) Stress: clean modulo known failures.
2) Libraries stress: pre-existing `op2.HasIconFlag()` assertion failures, what looks like #48798 and a bunch of Quic connection timeouts on Linux ARM. The timeouts were not present in the first stress run, so seems unlikely they're related.
3) GC stress: a number of timeouts, look to be pre-existing.

File: src/coreclr/jit/promotiondecomposition.cpp
Comment (SingleAccretion) [reply src/coreclr/jit/promotiondecomposition.cpp:744]: I do understand this to be correct, since:
1) Existing code already did this (reusing the destination/source nodes).
2) The store is added to the list (at most) once, and then all the "statements" are sequenced into a comma chain which replaces the original store in the tree.

Comment (jakobbotsch) [reply src/coreclr/jit/promotiondecomposition.cpp:744]: Yep, this looks fine to me. The original store is indeed replaced entirely by the result of the decomposition.

File: src/coreclr/jit/promotiondecomposition.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/promotiondecomposition.cpp:1003]: Move (or duplicate) the comment too?

File: src/coreclr/jit/promotion.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/promotion.cpp:566]: Whoops yes, feel free to remove that.

File: src/coreclr/jit/importer.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:3715]:
```suggestion
// impImportStaticFieldAddress: Generate an address of a static field
```

File: src/coreclr/jit/importer.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:3558]:
```suggestion
                unsigned structTempNum = lvaGrabTemp(true DEBUGARG("folding static readonly field empty struct"));
```
We spell out readonly below so might as well do it here too.

Comment (AndyAyersMS) [reply src/coreclr/jit/importer.cpp:3558]: Also there is one more "ro" just below here, can you change that one too?

File: src/coreclr/jit/importer.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:8832]: These aren't used until ~170 lines later, can we move this down closer to the use?

Comment (AndyAyersMS) [reply src/coreclr/jit/importer.cpp:8832]: In the store case (below) there are quite a few more uses, so if you want to leave things this way for symmetry then perhaps mark it as const so it is clearer  nothing in between is modifying it?

**Review by AndyAyersMS — APPROVED** [review]
I looked through everything and only have a few minor comment suggestions.

One broader nit is that we now use "value" and "data" somewhat interchangeably for what we used to sometimes call the RHS. I personally find "value" to be more semantically inspiring but don't feel like making a change here&mdash;assuming we decide we should&mdash;should hold up this PR.

Given the volume of the diffs it would be good to get a second reviewer. I would also like to hold off merging this until after the preview 5 snap tomorrow.

@jakobbotsch can you also do a review?

File: src/coreclr/jit/importer.cpp
Comment (SingleAccretion) [reply src/coreclr/jit/importer.cpp:8832]: The flags may get modified by the call `impImportStaticFieldAddress`, this happens when importing a static readonly `object` field. I can move both of these declarations closer to their uses if you think it would be better (the motivation for having this them this high up was indeed symmetry).

File: src/coreclr/jit/importer.cpp
Comment (SingleAccretion) [reply src/coreclr/jit/importer.cpp]: Addressed.

File: src/coreclr/jit/importer.cpp
Comment (SingleAccretion) [reply src/coreclr/jit/importer.cpp]: Addressed.

**Comment by SingleAccretion** [comment]
> One broader nit is that we now use "value" and "data" somewhat interchangeably for what we used to sometimes call the RHS. I personally find "value" to be more semantically inspiring.

Agree on that, there will be another "comment fixes" PR and I can fix `data`s then.

File: src/coreclr/jit/gentree.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/gentree.cpp:16424]: Shouldn't this be replaced by a `node->OperIsStore()` check?

Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:16424]: Ah, I suppose unlike ASG the stores are actually classified properly by `GenTree::IsValue`

File: src/coreclr/jit/lclmorph.cpp
Comment (jakobbotsch) [inline src/coreclr/jit/lclmorph.cpp:144]: Is `node->gtNext == nullptr` a meaningful check if `SequenceLocal` no longer unconditionally clears it?

**Review by jakobbotsch — APPROVED** [review]
LGTM also, great work!

File: src/coreclr/jit/promotion.cpp
Comment (SingleAccretion) [reply src/coreclr/jit/promotion.cpp:1331]: > though if we expect the flag to still be valid then we can leave it as is.

Right, the flag is still set (including on store nodes), so there is no need to change it now.

File: src/coreclr/jit/lclmorph.cpp
Comment (SingleAccretion) [reply src/coreclr/jit/lclmorph.cpp:144]: I had to recheck, but it turns out the answer is yes. The scenario is when we have half-consistent state during local morph with `m_stmtModified == true` and a new (bashed to) local address node inserted into the tree. Arguably, we should not be sequencing locals once `m_stmtModified` becomes `true`, but that is pre-existing behavior.

**Comment by jakobbotsch** [comment]
Thanks again! Awesome to have this in!

**Comment by am11** [comment]
Is there an issue tracking https://github.com/dotnet/runtime/issues/10873#issuecomment-412159692 (for the remainder of rationalizer phase)?

**Comment by SingleAccretion** [comment]
I don't believe there is; moving rationalization upwards means LIRizing HIR, which I don't believe there are plans for at this point in time.

**Pull request merged**
